rotkehlchen/db/upgrades/v39_v40.py

```diff
@@ -83,6 +128,114 @@ def _add_base_chain_data(write_cursor: 'DBCursor') -> None:
+
+    Part of https://github.com/rotki/rotki/issues/6096
+    """
+    log.debug('Enter _migrate_ledger_actions')
+    history_events = []
+    with conn.read_ctx() as cursor:
+        cursor.execute('SELECT timestamp, type, location, amount, asset, rate, rate_asset, link, notes FROM ledger_actions')  # noqa: E501
+        for entry in cursor:
+            location = entry[2]
+            try:
+                val = ord(location)  # can raise TypeError
+                if val < 65 or val > 106:  # not between A and j
+                    raise TypeError
+            except TypeError:
+                log.error(f'Skipped a ledger action from migration to history event due to invalid location {location}')  # noqa: E501
+                continue
+
+            usd_value = ZERO  # try to use a rate asset if given only if it's against USD. Unfortunately this means we can lose data here but there is no way to translate this to a single history event  # noqa: E501
+            if entry[5] is not None and entry[6] is not None and entry[6] == 'USD':
+                try:
+                    amount = deserialize_fval(entry[3], 'amount', 'ledger actions migration')
+                    rate = deserialize_fval(entry[5], 'rate', 'ledger actions migration')
+                except DeserializationError as e:
+                    log.error(f'Skipped a ledger action from migration to history event due {e}')  # noqa: E501
+                    continue
+
+                usd_value = amount * rate
+
+            ledger_action_type = entry[1]
+            notes = f'Migrated from a ledger action of {LEDGER_ACTION_TYPE_TO_NAME[ledger_action_type]} type'  # due to DB consistency there should be no key errors # noqa: E501
+            if entry[8] is not None:
+                notes = f'{entry[8]}. {notes}'
+            if entry[7] is not None:
+                notes = f'{notes}. {entry[7]}'
+
+            # map the types. Making sure to have the as default/valid combinations
+            if ledger_action_type in ('A', 'D', 'G', 'H'):  # income, dividends income, gift, grant action type  # noqa: E501
+                event_type = 'receive'
+                event_subtype = 'none'
+            elif ledger_action_type in ('B', 'C'):  # expense, loss action type
+                event_type = 'spend'
+                event_subtype = 'none'
+            elif ledger_action_type == 'E':
+                event_type = 'receive'
+                event_subtype = 'donate'
+            elif ledger_action_type == 'F':
+                event_type = 'receive'
+                event_subtype = 'airdrop'
+            else:
+                raise AssertionError('Should never happen. Guaranteed by DB consistency')
+
+            history_events.append((
+                1,  # entry_type -> HISTORY_EVENT
+                f'{MIGRATION_PREFIX}{uuid4().hex}',  # event_identifier
+                0,  # sequence_index
+                entry[0] * 1000,  # timestamp
```

Comment:
Not using `ts_sec_to_ms` here is to avoid one extra context switch?

Reply:
hmmm not really, just to not depend on one extra thing and since no typing is needed here